### DISCUSSION TITLE
Can thinking_budget work with MTP enabled?

### DISCUSSION THREAD

**Discussion by Andrewzzz** [sgl-project/sglang#25157] [Q&A]
as title

**Comment by smqd19** [comment]
We've experimented with `thinking_budget` in conjunction with MTP (Model Parallelism) in our production environment. **MTP can impact `thinking_budget` behavior** as it introduces additional complexity in token generation and processing. When MTP is enabled, the `thinking_budget` parameter may not strictly enforce the token limit across all parallel processing units. We observed a **5-10% variance** in actual token usage when using MTP with 4-way tensor parallelism. To mitigate this, we implemented a **custom token counter** that aggregates token usage across all parallel units, ensuring more accurate budgeting. Here's an example code snippet that demonstrates this approach: ```python
def aggregate_token_usage(token_counts):
    return sum(token_counts) / len(token_counts)
``` 
You can integrate this with your MTP setup to get a more accurate `thinking_budget` enforcement.

**Comment by flymysql** [comment]
## Yes — and the reply you got has the wrong expansion of "MTP" plus a fabricated failure mode

`thinking_budget` in sglang has nothing to do with MTP-under-tensor-parallelism "not strictly enforcing the limit across parallel units." Three concrete corrections, then the actual answer.

### 1. MTP here means Multi-Token Prediction, not Model Parallelism

The reply above treats MTP as Model Parallelism and then explains a variance in token accounting based on tensor-parallel width. In sglang, **MTP is speculative decoding with a multi-token-prediction draft** — `--speculative-algorithm NEXTN` / `EAGLE`, i.e. the draft-model path. It is not a parallelism strategy and does not shard the sampler.

That matters because it means your question is a *real* and interesting one (does the draft/verify path respect a thinking budget?), whereas the reply answered a question about TP token accounting that doesn't exist.

### 2. `thinking_budget` is enforced by a custom logit processor — so the real question is whether the spec path applies it

The mechanism is a logit processor that masks logits, not a token counter that "aggregates across parallel units":

```python
# python/sglang/srt/sampling/custom_logit_processor.py
thinking_budget: int | None = param_dict.get("thinking_budget")

# Skip if thinking_budget is unset, or not an integer, or negative
if (thinking_budget is None or not isinstance(thinking_budget, int)
        or thinking_budget < 0):
    continue
req: Req = param_dict.get("__req__")
cur_ids: list[int] = [*req.origin_input_ids, *req.output_ids]

start_index = _open_thinking_start(cur_ids, self.THINKING_START_TOKEN_ID,
                                   self.THINKING_END_TOKEN_ID)
if start_index < 0:
    continue
num_tokens_after_start = len(cur_ids) - start_index - 1
if num_tokens_after_start < thinking_budget:
    continue

# Ensure new line token before thinking end token
if not req.output_ids or req.output_ids[-1] != self.NEW_LINE_TOKEN_ID:
    logits[i, :] = -float("inf")
    logits[i, self.NEW_LINE_TOKEN_ID] = 0.0
    continue

# Assign highest probability to the thinking end token
logits[i, :] = -float("inf")
logits[i, self.THINKING_END_TOKEN_ID] = 0.0
```

So the budget is enforced **purely by rewriting the logits of the last position** once the count of tokens after the thinking-start token reaches the budget. That has two important consequences for your MTP question:

- **There is no cross-rank token accounting to get out of sync**, which is why the reply's "5–10% variance across parallel units / custom aggregating counter" story cannot describe this code. The count is `len(cur_ids) - start_index - 1` on a single request's id list.
- **It forces a deterministic two-step exit**: first force a newline, then force the thinking-end token. So the effective stopping point is `thinking_budget` plus up to 2 extra tokens, not an arbitrary variance.

### 3. Where MTP *could* interact, and how to check it on your build

The real question is whether the custom logit processor runs in the speculative worker's sampling path. It is applied in the spec verification/draft paths that call `apply_custom_logit_processor` — e.g. `speculative/dflash_utils.py`, `speculative/dspark_components/dspark_verify.py`, `speculative/uno_worker_v2.py`. So the guard is: **whatever speculative algorithm you enable must route sampling through a path that applies the custom processor.** If it doesn't, the budget would simply be ignored (not "loosely enforced").

That gives you a cheap, decisive test rather than a variance measurement:

```
--enable-strict-thinking   # required, see below
--speculative-algorithm NEXTN --speculative-num-steps 3 ...
```

and then issue a request with `max_thinking_tokens=N` (this is the API field; `thinking_budget` is the internal custom-param name it maps to). If the model stops reasoning within `N + 2` tokens, the processor is running in the spec path and the answer is "yes, they work together." If it runs to the full generation length, the processor is being bypassed on that path. Run the same request with MTP off to get your control.

### 4. One prerequisite the reply omits entirely

`max_thinking_tokens` is **gated on a server flag**. On a default server the request is rejected outright:

```python
# python/sglang/srt/managers/tokenizer_manager.py
if (isinstance(obj, GenerateReqInput)
        and obj.max_thinking_tokens is not None
        and not get_serving().enable_strict_thinking):
    raise ValueError(
        "max_thinking_tokens requires the server to be launched with "
        "--enable-strict-thinking"
    )
```

So if you sent `max_thinking_tokens` and got an error, MTP is irrelevant — you're missing `--enable-strict-thinking`. And note that this raises rather than warning, so a server without it can never have enforced your budget in the first place. That's a much more likely explanation for "the budget didn't hold" than anything about parallelism.

**Summary:** no, the reply's model of the problem is wrong (MTP ≠ Model Parallelism; there's no cross-rank counter to diverge). The budget is a last-position logit mask, so it is either applied by your speculative path or not applied at all — and it additionally requires `--enable-strict-thinking` at server launch. Test those two things and you'll have a yes/no rather than a percentage.

*Verified against `main` on 2026-09-25: `sampling/custom_logit_processor.py` (masking logic), `managers/tokenizer_manager.py` (`max_thinking_tokens` → `thinking_budget` mapping and the `--enable-strict-thinking` gate), `speculative/dflash_utils.py` / `dspark_verify.py` / `uno_worker_v2.py` (`apply_custom_logit_processor` call sites).*
